training/tests/integration/test_benchmark.py

```diff
@@ -0,0 +1,554 @@
+            )
+        row = maybeRow
+
+        assert row["testName"] == benchmarkName  # sanity check, should always pass
+        benchmarkValue = BenchmarkValue(
+            name=benchmarkName, value=row["value"], unit=row["unit"], date=row["date"], commit=row["commit"],
+        )
+        LOGGER.debug(benchmarkValue)
+        # update dict of results
+        self.benchmarkValues[benchmarkValue.name] = benchmarkValue
+
+    def getValues(self, names: list[str]):
+        for name in names:
+            self.getValue(name)
+
+    # Tests a given benchmark result against what is found on the server
+    def compare(self, localValue: BenchmarkValue, failOnMiss=False):
+        # check if the server has a reference value
+        referenceValue = self.getValue(localValue.name)
+
+        if referenceValue is None:
+            if failOnMiss:
+                print(f"Benchmark server does not contain a measurement for {localValue.name}")
+                return False
+            print(f"{localValue.name} not found on server. Passing anyway because 'failOnMiss=False'")
+            return True
+        LOGGER.debug("didnt pass")
+        passed = False
+
+        comp = localValue.op
+        refVal = referenceValue.value
+        localVal = localValue.value
+        tolerance = localValue.tolerance
+
+        # Sanity checking that benchmark metadata matches
+        # assert localValue.op == referenceValue.op #wont work, need to pass some inputs
+        assert localValue.unit == referenceValue.unit
+
+        percent_diff = 1 - (refVal / localVal)
+        passedWithinTolerance = False
+        if comp(percent_diff, 0):
+            passed = True
+        # didnt pass straight away, try pass within tolerance
+        elif tolerance != 0 and tolerance / 100 >= abs(percent_diff):
+            passed = True
+            passedWithinTolerance = True
+        else:
+            passed = False
+
+        result_str = ""
+        if passed:
+            if passedWithinTolerance:
+                result_str += f"PASS. Local value for {localValue.name} is within {tolerance}% tolerance of the reference value "
+            else:
+                result_str += (
+                    f"PASS. Local value for {localValue.name} has improved compared to the reference value "
+                )
+
+        else:
+            result_str += f"FAIL. Local value for {localValue.name} has degraded compared to the reference value "
+        result_str += f"({localVal:.2f}{localValue.unit} local vs {refVal:.2f}{referenceValue.unit} reference)"
+        print(result_str)
+
+        return passed
+
+    # trys to update a metric on a remote server, with a given benchmarkValue
+    #if overwrite is true, setValue wont try append. it will be like the exisitng file doesnt exist
+    def setValue(self, value: BenchmarkValue, overwrite=False):
```

Comment:
If we run this from many gpus of do we control that this commands just get executed by just on the processess?